File: tang/src/main/java/com/microsoft/tang/annotations/NamedParameter.java
package com.microsoft.tang.annotations;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Target(ElementType.TYPE)
@Documented
@Retention(RetentionPolicy.RUNTIME)
public @interface NamedParameter {
  //Class<?> type() default String.class;
  String doc() default "";
  String short_name() default "";
  // One of the following should be set.
  String default_value() default "";
  Class<?> default_class() default Void.class;
  String[] default_values() default {};
  Class<?>[] default_classes() default {};
}

File: tang/src/main/cs/Tang/Annotations/NamedParameter.cs
using System;

namespace Com.Microsoft.Tang.Annotations
{
    [System.AttributeUsage(System.AttributeTargets.Class)]
    public class NamedParameterAttribute : System.Attribute
    {
        public string Documentation { get; set; }
        public string ShortName { get; set; }
        public string DefaultValue { get; set; }
        public Type DefaultClass { get; set; }
        public string[] DefaultValues { get; set; }
        public Type[] DefaultClasses { get; set; }

        public NamedParameterAttribute(string documentation = "", string shortName = "",
            string defaultValue = "", Type defaultClass = null, string[] defaultValues = null, Type[] defaultClasses = null)
        {
            this.Documentation = documentation;
            this.ShortName = shortName;
            this.DefaultValue = defaultValue;
            this.DefaultClass = defaultClass;
            this.DefaultValues = defaultValues;
            this.DefaultClasses = defaultClasses;
        }
    }
}
